controllers/oauth2client_controller_test.go

```diff
@@ -0,0 +1,227 @@
+package controllers
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"github.com/ory/hydra-maester/api/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/square/go-jose.v2/json"
+)
+
+func TestNewRequest(t *testing.T) {
+
+	r := OAuth2ClientReconciler{
+		HydraURL: url.URL{
+			Path:   "example.com",
+			Scheme: "http",
+		}}
+
+	assert := assert.New(t)
+
+	for desc, tc := range map[string]struct {
+		method  string
+		relPath string
+		body    interface{}
+	}{
+		"Basic GET request": {
+			method:  http.MethodGet,
+			relPath: "some-endpoint",
+			body:    nil,
+		},
+		"Basic POST request": {
+			method:  http.MethodPost,
+			relPath: "",
+			body:    v1alpha1.OAuth2ClientJSON{Name: "some_name", GrantTypes: []string{"type1"}, Scope: "some,scope"},
+		},
+	} {
+		t.Run(fmt.Sprintf("case/%s", desc), func(t *testing.T) {
+
+			//when
+			req, err := r.newRequest(tc.method, tc.relPath, tc.body)
+
+			//then
+			require.NoError(t, err)
+			assert.Equal(tc.method, req.Method)
+			if tc.relPath == "" {
+				assert.Equal(r.HydraURL.String(), req.URL.String())
+			} else {
+				assert.Equal(fmt.Sprintf("%s/%s", r.HydraURL.String(), tc.relPath), req.URL.String())
+			}
+
+			require.NotEmpty(t, req.Header.Get("Accept"))
+			assert.Equal("application/json", req.Header.Get("Accept"))
+
+			if tc.body != nil {
+				require.NotEmpty(t, req.Header.Get("Content-Type"))
+				assert.Equal("application/json", req.Header.Get("Content-Type"))
+
+				buf := new(bytes.Buffer)
+				_, _ = buf.ReadFrom(req.Body)
+
+				var c v1alpha1.OAuth2ClientJSON
+				err = json.Unmarshal(buf.Bytes(), &c)
+
+				assert.Equal(c, tc.body)
+			}
+		})
+	}
+}
+
+const (
+	testID            = "test-id"
+	schemeHTTP        = "http"
+	testClient        = `{"client_id":"test-id","client_name":"test-name","scope":"some,scopes","grant_types":["type1"]}`
+	testClientCreated = `{"client_id":"test-id-2", "client_secret": "TmGkvcY7k526","client_name":"test-name-2","scope":"some,other,scopes","grant_types":["type2"]}`
+	emptyBody         = `{}`
+)
+
+var testOAuthJSON = &v1alpha1.OAuth2ClientJSON{
+	Name:       "test-name-2",
+	Scope:      "some,other,scopes",
+	GrantTypes: []string{"type2"},
+}
+
+func TestCRUD(t *testing.T) {
```

Comment:
Contrary to the name, this test does not test the controller but the rest client. It should be moved along with the client to the "hydra" package

Reply:
yes sir!